libs/agent-sm/agent-policy/src/main/java/org/opensearch/secure_sm/policy/PolicyFile.java

```diff
@@ -0,0 +1,346 @@
+            String path = url.getFile().replace('/', File.separatorChar);
+            path = URLDecoder.decode(path, StandardCharsets.UTF_8);
+            return new FileInputStream(path);
+        } else {
+            return url.openStream();
+        }
+    }
+
+    private CodeSource getCodeSource(GrantEntry grantEntry, PolicyInfo newInfo) throws PolicyInitializationException {
+        try {
+            Certificate[] certs = null;
+            URL location = (grantEntry.codeBase != null) ? newURL(grantEntry.codeBase) : null;
+            return canonicalizeCodebase(new CodeSource(location, certs));
+        } catch (Exception e) {
+            throw new PolicyInitializationException("Failed to get CodeSource", e);
+        }
+    }
+
+    private void addGrantEntry(GrantEntry grantEntry, PolicyInfo newInfo) throws PolicyInitializationException {
+        CodeSource codesource = getCodeSource(grantEntry, newInfo);
+        if (codesource == null) {
+            throw new PolicyInitializationException("Null CodeSource for: " + grantEntry.codeBase);
+        }
+
+        PolicyEntry entry = new PolicyEntry(codesource);
+        List<PermissionEntry> permissionList = grantEntry.permissionElements();
+        for (PermissionEntry pe : permissionList) {
+            expandPermissionName(pe);
+            try {
+                Optional<Permission> perm = getInstance(pe.permission, pe.name, pe.action);
+                if (perm.isPresent()) {
+                    entry.add(perm.get());
+                }
+            } catch (ClassNotFoundException e) {
+                // these were mostly custom permission classes added for security
+                // manager. Since security manager is deprecated, we can skip these
+                // permissions classes.
+                if (PERM_CLASSES_TO_SKIP.contains(pe.permission)) {
+                    continue; // skip this permission
+                }
+                throw new PolicyInitializationException("Permission class not found: " + pe.permission, e);
+            }
+        }
+        newInfo.policyEntries.add(entry);
+    }
+
+    private void expandPermissionName(PermissionEntry pe) {
+        if (pe.name == null || !pe.name.contains("${{")) {
+            return;
+        }
+
+        int startIndex = 0;
+        int b, e;
+        StringBuilder sb = new StringBuilder();
+
+        while ((b = pe.name.indexOf("${{", startIndex)) != -1 && (e = pe.name.indexOf("}}", b)) != -1) {
+            sb.append(pe.name, startIndex, b);
+            String value = pe.name.substring(b + 3, e);
+            sb.append("${{").append(value).append("}}");
+            startIndex = e + 2;
+        }
+
+        sb.append(pe.name.substring(startIndex));
+        pe.name = sb.toString();
+    }
+
+    private static final Optional<Permission> getInstance(String type, String name, String actions) throws ClassNotFoundException {
+        Class<?> pc = Class.forName(type, false, null);
```

Comment:
Since we know what classes the agent will support up front, should we actually instantiate these permissions or skip the unsupported one's while parsing?

Reply:
yeah, that's a good idea.

Reply:
i gave a shot at it [here](https://github.com/opensearch-project/OpenSearch/pull/17712#issuecomment-2779681591) and it broke a lot of tests .  I'll need some time to debug this but would like to go ahead with current changes to keep the bundling and distribution unblocked.

Reply:
I took a second look and the logic you have in this PR only tries to instantiate known permissions. At first glance I thought it was trying to instantiate any permission entry in the policy file similar to https://github.com/openjdk/jdk17/blob/master/src/java.base/share/classes/sun/security/provider/PolicyFile.java#L849-L884. This change looks good. Marking this thread as resolved.